Earthfile

```diff
@@ -0,0 +1,290 @@
+
+  # test in native rust, not in wasm for a future rust SDK
+  # without --tests, the --doc is causing a link error "syntax error in VERSION script"
+  RUN bash -c '\
+    cargo test --locked --package typegraph_core --tests;'
+
+test-meta-cli:
+  FROM +pretest-rust
+
+  RUN bash -c '\
+    cargo run --profile $CARGO_PROFILE --locked --package meta-cli -- --help; \
+    cargo test --locked --package meta-cli; '
+
+test-libs:
+  FROM +pretest-rust
+
+  # from old test-libs
+  RUN bash -c '\
+    cargo test --locked --exclude meta-cli --exclude typegate \
+      --exclude typegraph_engine --exclude typegraph_core --workspace;'
+
+img-test-typegate:
+  FROM +test-meta-cli
+
+  RUN bash -c '\
+    cargo build --profile $CARGO_PROFILE -p meta-cli; \
+    cargo build --profile $CARGO_PROFILE -p xtask; ' # xtask and meta-cli are used by the test suite
+
+  COPY --dir dev/ .
+
+  RUN bash -c '\
+    deno run -A dev/update.ts --cache-only || deno run -A dev/update.ts --cache-only;'
+
+  ENTRYPOINT ["bash", "-c", "\
+    deno run -A dev/test.ts --threads 2 -- --coverage=coverage; \
+    deno --unstable coverage ./coverage --lcov > coverage.lcov; "]
+  SAVE IMAGE test-typegate:latest
+
+test-typegate:
+  FROM earthly/dind:alpine-3.18-docker-23.0.6-r4
+  WORKDIR /app
+  COPY dev/envs .
+  WITH DOCKER \
+          --compose compose.base.yml \
+          --compose compose.prisma.yml \
+          --compose compose.s3.yml \
+          --load test-typegate:latest=+img-test-typegate
+      RUN docker run test-typegate:latest
+  END
+
+
+build-typegate:
+  FROM +chef-cook
+
+  DO +COPY_SOURCES_RUST
+  RUN cargo build --profile $CARGO_PROFILE --package typegate
+
+  COPY +deno-bin/deno /bin/deno
+  COPY --dir dev/ .
+  RUN deno run -A dev/update.ts --cache-only --src-only && mkdir -p typegate/tmp
+
+  SAVE ARTIFACT target/$CARGO_PROFILE/typegate
+  SAVE ARTIFACT typegate/deno.lock
+  SAVE ARTIFACT typegate/tmp
+  # NOTE: this must match with DENO_DIR
+  SAVE ARTIFACT /deno-dir/
+
+tini:
+  FROM +rust-builder
+  ARG TINI_VERSION=v0.19.0
+
+  RUN curl -fsS \
+    https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini-${TARGETARCH} \
+    -o /tini; \
+    chmod +x /tini; \
+    mkdir -p /lib/sym; \
+    ln -s /lib/aarch64-linux-gnu /lib/sym/arm64; \
+    ln -s /lib/x86_64-linux-gnu /lib/sym/amd64;
+
+  SAVE ARTIFACT /tini
+  SAVE ARTIFACT /lib/sym
+
+img-typegate:
+  FROM gcr.io/distroless/cc-debian11:${DISTROLESS_TAG}
+
+  ENV NO_COLOR true
+
+  COPY +tini/tini /tini
+  COPY +tini/sym /lib/sym
+
+  COPY +build-typegate/root/.wasmedge /nonroot/.wasmedge
+  COPY +build-typegate/lib/*-linux-gnu/libz.so* /lib/sym/${TARGETARCH}
+  COPY +build-typegate/target/$CARGO_PROFILE/typegate /bin/typegate
+
+  # the typegate ecma sources 
+  COPY typegate/deno.jsonc typegate/import_map.json ./typegate/
+  COPY typegate/engine/*.js typegate/engine/*.ts ./typegate/engine/
+  COPY typegate/src ./typegate/src/
+  COPY dev/LICENSE-Elastic-2.0.md LICENSE.md
+
```

Comment:
earthly is awesome! imo, there is 3 main drawbacks :
- for task, there is a 3-4s to init the process (even warm) which make it inconvenient to use it for developper task
- there is no [watch mode](https://github.com/earthly/earthly/issues/1052), however this should not be that complex for our use case (for the generic case, it is indeed). Its main competitor [Dagger](https://docs.dagger.io/sdk/python/628797/get-started) exposes the low-level api which is nothing else that the [LLB](https://github.com/moby/buildkit#exploring-llb) of buildkit which allows us to combine it with a custom watcher
- when onboarding the project installing earthly + running on the docker backends is of a heavier footprint than running something like jkjk, especially with people without cloud env

Reply:
as we already manipulate the env with jkjk, I would feel it as convient to add more logic into it similar to [taskfile](https://taskfile.dev) and enabling whiz.yaml to offload most of the task execution to jkjk (watch, dependencies). This would allow :
- run ghjk for single task in dev
- run ghjk for pre-commit hooks and ensuring the hook is configured
- run ghjk for tests
- run ghjk for continuous integration
- (later) run ghjk to build docker like earthly/dagger with llb by starting from scratch and simply brining in all the binary needed for execution

Reply:
Yeah, I've removed the Earthfile for the points listed here and more issues.  The artifacts for metatype builds are too large and it keeps bugging out trying to make some of those layers.